Add guide on how to connect local nodes

OrfeasLitos opened the pull request:
Demo of two regtest nodes connecting. Resolves #106.

pinheadmz commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,95 @@
+
+  // no peers for the spvNode yet :(
+  console.log('spvNode\'s peers before connection:', spvNode.pool.peers.head());
+  assert.equal(spvNode.pool.peers.head(), null);
```
When I ran this script, I actually already had a peer:
`spvNode's peers before connection: <Peer: handshake=false host=18.191.95.172:48445 outbound=true ping=-1>`

...and so the assert failed. (That's the IP address of my own server). The "before connection" printout is interesting, but I don't know if we need the `assert` there... unless there is some other way to definitively clear all peers before starting.

pinheadmz commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,95 @@
+  spvNode.pool.peers.add(peer);
+
+  // allow some time to establish connection
+  await delay(3000);
```
I actually had to change this delay to 10 seconds `delay(10000)` on my machine or the second assert would fail. I wonder if we could use an await/async method here instead of gambling on how long the connection will take? I haven't tested this yet but maybe `peer.open()`? https://github.com/bcoin-org/bcoin/blob/master/lib/net/peer.js#L347-L361

pinheadmz commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,95 @@
+  // nodes are now connected!
+  console.log('spvNode\'s peers after connection:', spvNode.pool.peers.head());
+  assert.equal(spvNode.pool.peers.head().inspect(),
+    '<Peer: handshake=true host=127.0.0.1:48445 outbound=true ping=-1>');
```
Once I adjusted the timeout above and commented out the first assert, this assert succeeded. But still, again, I think it is a long string to assert equality. See below for my idea on how to wrap this up nicely.

OrfeasLitos commented:
Thanks for the constructive suggestions. These would make the example much more robust and informative. Avoiding timeouts and `assert`s is the ideal case. I managed to avoid the second timeout with a less-than-ideal trick, but I'm not sure how to avoid the first timeout. Specifically, the spvNode tries to connect to a default port (`48444`) and if I try to explicitly disconnect it from this port there is an error. I couldn't find some way to ensure that the peer list is empty (apart from the obviously ugly busy-wait). Do you have any suggestions?

OrfeasLitos commented:
Another question. It seems that the example you gave with the `generateToAddress()` does not work. On the wire I can just see a `get_block` inv packet, not a `get_tx`. The other solution would be to actually create `wallet`s just for the demo, but first I find this an overkill and second there is currently a bug in bcoin where if an spv node creates a walletdb, it overwrites its spv filter with the tx filter and the spv node can't connect anymore (https://github.com/bcoin-org/bcoin/pull/578). So the guide would not work until the next release. Any alternative ideas for the demo? Or a workaround to the `get_block` inv?

OrfeasLitos commented:
I like your approach, with the logging and `maxOutbound: 1`. I removed the `assert()`s (and stopped `require()`ing the library) and fixed a couple of typos. I realized that `createOutbound()` actually does all the connection plumbing, except for actually adding the peer to the `PeerList` (what `spvNode.pool.peers.add(peer)` does, c.f. https://github.com/bcoin-org/bcoin/blob/1bde15b9a24c2d1c71db33864b326b58ae5d2e60/lib/net/pool.js#L2984-L2986, https://github.com/bcoin-org/bcoin/blob/1bde15b9a24c2d1c71db33864b326b58ae5d2e60/lib/net/pool.js#L1054-L1066). Thus `createOutbound()` should go after `delay(800)` and together with `peers.add()`. In practice, this doesn't change the output, but let's change it for the sake of clarity.

I used the tool `p-event` to `await` for the emission of a `'peer connect'` event from `spvNode.pool` and the `'block'` event from `spvNode`. This way we avoid the long, arbitrary delay of 10s (at the cost of continuing before the handshake is complete) and we can also cleanly close the nodes and allow the script to exit.

In slightly older commits, the SPV node used to correctly receive only one transaction! The regression happens here: https://github.com/bcoin-org/bcoin/commit/a0ac953079a5ce09eb23828239ed5ab70b975f0f . If it is fixed, we can simply send the two txs and avoid the blocks. This solution seems better because the difference between the two blocks as printed is too small.

pinheadmz commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,153 @@
+  logConsole: false,
+  logLevel: 'spam',
+
+  // reduce log spam on SPV node (won't warn on Full node)
```
sorry this is my typo, it should say `won't work on full node` -- the full node will still try to connect to that 1 outbound peer and the log will fill up anyway. `maxOutbound: 0` returns an error, I tried it :-)

OrfeasLitos replied:
Ah, now I see what you meant. Maybe we can achieve the intended behaviour by making the full node `selfish`. I'll try it and add it if it reduces logging but still broadcasts stuff (I suppose it should).

OrfeasLitos replied:
I will add a comment explaining why `selfish` and why users shouldn't do it in general :)

pinheadmz replied:
I thought `selfish` prevented all blocks and txs from going out...? My intent with the `maxOutbound` was just to reduce the `attempt to connect to peer` stuff in the log... it's not that crucial either way :-)

OrfeasLitos replied:
Apparently `selfish` mode doesn't work. First of all, selfish nodes do not connect to spv nodes. Second, even if I try to undo some of the damage done by the `selfish` flag, apparently the handshake fails with the (selfish) full node never sending back the `verack` packet... Probably what I tried to do is unsupported, but still it's weird behavior. Anyway, let's forget the `selfish` trick.

pinheadmz commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,153 @@
+  });
+  spvNode.on('tx', (tx) => {
+    console.log('-- SPV node received tx: --\n', tx);
+  });
```
Can't get `'tx'` event to fire, let's just pull these lines out (I'm going to continue to investigate why we can't catch this event, maybe we can update this guide once I figure it out!)

pinheadmz commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,153 @@
+  //const tx2 = bcoin.TX.fromRaw('010000000106b014e37704109fefe2c5c9f4227d68840c3497fc89a9832db8504df039a6c7000000006a47304402207dc8173fbd7d23c3950aaf91b1bc78c0ed9bf910d47a977b24a8478a91b28e69022024860f942a16bc67ec54884e338b5b87f4a9518a80f9402564061a3649019319012103cb25dc2929ea58675113e60f4c08d084904189ab44a9a142179684c6cdd8d46affffffff0280c3c901000000001976a91400ba915c3d18907b79e6cfcd8b9fdf69edc ...
+  //fullNode.sendTX(tx2);
+  // await for the second transaction to be received by the SPV node
+  //await pEvent(spvNode, 'tx');
```
All these commented out lines, let's just remove them for clarity

pinheadmz commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,153 @@
+gracefully closed.
+
+The script should work out of the box, the only requirement is having `bcoin`
+installed.
```
I had to install `p-event` on my system as well :-) 
Can you add a little instruction for that? Maybe a one-line description of the package as well?

OrfeasLitos replied:
Oops, I forgot to change the prose as well.

pinheadmz commented:
Ah ok, I just tested by re-installing bcoin at https://github.com/bcoin-org/bcoin/commit/5b56323310c36620667ef347ad76615ddfad00c7

and you're right, the TX fires. Let me see if we can get a fix pulled and then maybe yeah, use tx events instead of the block

OrfeasLitos commented:
I will try to see if the example works with txs on the latest `npm` release. If it works (which is what I expect) I will remove the block creation and just send 2 txs. I assume that guides should work on the latest version, not on the latest commit, so it's fine if it doesn't work until https://github.com/bcoin-org/bcoin/pull/598 is merged.

bucko13 commented on guides-markdown/connect-local-nodes.md:
````diff
@@ -0,0 +1,141 @@
+Start and manually connect two regtest nodes in a single script.
+```
+
+When we start two nodes in the regtest (loopback network, 127.0.0.1) with
````
"on the regtest network" instead of "in the regtest"

bucko13 commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,141 @@
+When we start two nodes in the regtest (loopback network, 127.0.0.1) with
+manually set ports, they don't get connected automatically because they don't
+know each other's port. We thus have to connect them manually. In this example
+we will fire up one SPV and one Full node and initiate a connection from the
```
I think either "Full Node" or "full node". SPV is capitalized because it's an acronym, but "Full" isn't a proper noun on its own

bucko13 commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,141 @@
+manually set ports, they don't get connected automatically because they don't
+know each other's port. We thus have to connect them manually. In this example
+we will fire up one SPV and one Full node and initiate a connection from the
+SPV side. The SPV node is given the address of the Full node and attempts to
```
same here.

bucko13 commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,141 @@
+know each other's port. We thus have to connect them manually. In this example
+we will fire up one SPV and one Full node and initiate a connection from the
+SPV side. The SPV node is given the address of the Full node and attempts to
+connect. If everything goes well, the two nodes connect. The Full node then
```
Same here. Also, let's split the paragraph here when you start talking about transactions (easier to read smaller paragraphs and this is kind of another section). So "Next, we'll have the full node broadcast two transactions."

OrfeasLitos replied:
(+1) I also rephrased the first sentence of the newly split paragraph to "To demonstrate that the connection was successful, we'll have full node..." to make clear why on earth we deal with txs.

bucko13 commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,141 @@
+by the SPV node and the second isn't. Finally the two nodes are gracefully
+closed.
+
+We use the library `p-event` which promisifies events in order to `await` for
```
Cool!

bucko13 commented on guides-markdown/connect-local-nodes.md:
```diff
@@ -0,0 +1,141 @@
+  httpPort: 48449, // avoid clash of ports
+
+  // write log file and chain data to specific directory
+  prefix: '~/connect-test/SPV',
```
small nit, but it might be nice to make `~/connect-test` a constant since it's used in multiple places. This is nice in case someone forgets to change both and ends up with two different project directories. Also if you use node's `os.homedir()` method instead of `~` it can work across platforms since I don't think Windows recognizes `~`.

OrfeasLitos replied:
Doesn't windows use backslashes instead of slashes as well? I don't have a windows box to test upon, but I think something in the lines of https://gist.github.com/domenic/2790533 should be used. More info: https://nodejs.org/docs/latest/api/path.html. I think I got it correctly; At least my box is happy. If anyone can test it on a windows box, please tell me if it works for them.

bucko13 replied:
Well, the beauty if using Node's `os` module is that it will take care of environment specific conditions for you so you don't even have to worry about it.

bucko13 replied:
i.e. `const prefix = `${require('os').homedir()}/connect-test/SPV`
if you look at the bcfg library which manages this in bcoin, it makes heavy use of this convention, even explicitly replacing instances of `~` passed by the user.

OrfeasLitos replied:
Thanks for pointing me there, I found `path.join()` which accomplishes what I wanted more simply than `path.format()`. Let me know if you'd like any more changes.

bucko13 commented:
Resolves #106
